ubuntu 18.04 dockerfile

zxkjack123 opened the pull request:
This PR adds a dockerfile for mcnp2cad cli on ubuntu 18.04.

But the cli is not working well yet, with the following problem when convert the `tests/sample1`:
```
Saving file "out.sat"...			iGeom error (22): saving the output file out.sat
 * Unknown geometry file extension and no file type.
 done.
LineExtractor out of lines, cannot takeLine().
```

zxkjack123 commented:
@makeclean You told me about changing one of the header files of mcnp2cad, to set the default CAD filename. But I not not sure with one to change, could you please repeat the change you made use this docker image?

gonuke commented:
The default name is in `options.hpp`. The ideal solution is make this a compile time option and/or auto generated by CMake, similar.

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN apt-get update -y && \
+    apt-get upgrade -y && \
+    apt-get install -y software-properties-common build-essential && \
+    apt-get install -y wget git vim && \
```
why not installing everything in a single line ?

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+    apt-get upgrade -y && \
+    apt-get install -y software-properties-common build-essential && \
+    apt-get install -y wget git vim && \
+    add-apt-repository ppa:jonathonf/gcc-9.1 && \
```
same here you could add the repo before hand and then install everything in a single time.

bam241 replied:
Also is gcc-9 required ?
I don't know what is the purpose of this docker container, but if it is for CI-testing maybe staying with the normal gcc version is better as it will be what most of the user have....

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+    mkdir build && cd build  && \
+    cmake -DOCE_INSTALL_PREFIX:PATH=$HOME/opt/OCE-0.18.3 .. && \
+    make  -j4 && make install/strip
+RUN cd $HOME && \
```
Is a new `RUN` command really required ?

gonuke replied:
I think it is good to have a different `RUN` command her because it creates a snapshot for incremental steps.  Installing OCE is a semantically separate step.

bam241 replied:
I meant for the cleaning part, should it not be part of the installation process ?

gonuke replied:
Oops - yes, I agree, I was looking at the wrong scope of lines

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+    rm ./OCE-0.18.3.tar.gz  && rm -rf ./oce-OCE-0.18.3
+## OCE evn
+RUN echo 'export OCE_DIR=$HOME/opt/OCE-0.18.3' >> $HOME/.bashrc
+RUN echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$OCE_DIR/lib' >> $HOME/.bashrc
```
Same for those `echo`  maybe a single `RUN` is enough ?
maybe you could tied this to the install `RUN` ? 

Also maybe you should consider using the ENV command?

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN echo 'export PATH=$PATH:$HOME/opt/CGM/bin' >> $HOME/.bashrc
+RUN echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/opt/CGM/lib' >> $HOME/.bashrc
+RUN echo 'export C_INCLUDE_PATH=$C_INCLUDE_PATH:$HOME/opt/CGM/include' >> $HOME/.bashrc
+RUN echo 'export CPLUS_INCLUDE_PATH=$CPLUS_INCLUDE_PATH:$HOME/opt/CGM/include' >> $HOME/.bashrc
```
Same idea here, why not using `ENV` command, i.e. why this need to be in the `.bashrc` ?

zxkjack123 replied:
All these exports to `~/.bashrc` is not required in the install process. But it would be convenient for tests and running docker after installing. So, I put them here.

There are two `ENV` commands before installing `mcnp2cad`. These two are required and I used the `ENV`.

bam241 replied:
yes but what I meant is those environment variable you export in the .bashrc, will only be available if you manually `source .bashrc` or run bash in your docker container.
I believe if you use ENV, they will be defined regardless of the way you use the container... (specially if you plan to use this in CI)

zxkjack123 replied:
Got it, thanks @bam241.

bam241 replied:
![](https://d33wubrfki0l68.cloudfront.net/8d799fc311d166c3a7d9f1a7970e1dd1d43bf141/0a855/images/docker-env-vars/docker_environment_build_args.png)

from [[here](https://vsupalov.com/docker-arg-env-variable-guide/#arg-and-env-availability)]

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+ENV LD_LIBRARY_PATH $HOME/opt/CGM/lib:$HOME/opt/OCE-0.18.3/lib
+ENV CPLUS_INCLUDE_PATH $HOME/opt/CGM/include
+RUN git clone https://github.com/svalinn/mcnp2cad.git $HOME/opt/mcnp2cad
+RUN cd $HOME/opt/mcnp2cad  && mkdir build && cd build && \
```
for readability I would suggest to have a command per line (keeping everything in a single `RUN` command)

bam241 replied:
also multiple `RUN` are probably not required....

bam241 commented on docker/dockerfile:
```diff
@@ -0,0 +1,58 @@
+RUN echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/opt/CGM/lib' >> $HOME/.bashrc
+RUN echo 'export C_INCLUDE_PATH=$C_INCLUDE_PATH:$HOME/opt/CGM/include' >> $HOME/.bashrc
+RUN echo 'export CPLUS_INCLUDE_PATH=$CPLUS_INCLUDE_PATH:$HOME/opt/CGM/include' >> $HOME/.bashrc
+RUN cd $HOME/opt/CGM/include && cp ./cgm/*.h* .
```
this could be added at the end of the install `RUN` command.

zxkjack123 commented:
Update: changed according to the suggestion of @bam241:
- Remove `gcc-9.1` requirement,
- Put some commands in one `RUN` block but multiple line,
- Remove exports to `~/.bashrc`, replaced by `ENV`.

bam241 commented:
nice work! LGTM!

zxkjack123 commented:
> The default name is in `options.hpp`. The ideal solution is make this a compile time option and/or auto generated by CMake, similar.

@gonuke I think we should make another PR to resolve the problem.

gonuke commented:
> > The default name is in `options.hpp`. The ideal solution is make this a compile time option and/or auto generated by CMake, similar.
> 
> @gonuke I think we should make another PR to resolve the problem.

No problem making it a different PR - I was just responding to your question to @makeclean

gonuke commented:
Thanks @zxkjack123
